### PR TITLE
N-D Finite Differences

### PR DESCRIPTION
**Pull request by StevenWhitaker**
I decided to take a stab at implementing N-D finite differences. It looks like the N-D version is even faster than the current 2D version (and the relative speedup increases as the dimensionality increases)! And no `CartesianIndices` necessary :)

Let me know if you want me to change anything.

### CODE REVIEW

File: src/regularize/diffs.jl [hunk 1]
```diff
@@ -1,13 +1,14 @@
 #=
 diffs.jl
 finite differences
-2019-03-96 Jeff Fessler, University of Michigan
+2019-03-06 Jeff Fessler, University of Michigan
```
Comment (StevenWhitaker) [inline src/regularize/diffs.jl:4]: I'm guessing you meant 06, but I guess it could also be 16 or 26.

File: src/regularize/diffs.jl [hunk 2]
```diff
@@ -1,99 +1,122 @@
...
-	return out2d ? z : z[:] # array or vector?
+function diffnd_adj(d::AbstractVector{<:Number}, N::Vararg{Int,D}
+    ; dims=1:D, outnd=false) where {D}
+
+# todo: diff2d_adj etc. 1-liners for backward compat
+# todo: N::Dims instead of Vararg?
+# todo: remove outnd option for type inference?
+# todo: discuss, >1 should be needed only for the "dims" dimensions?
+    # Note that N must be strictly greater than 1 for each dimension specified
+    # by dims, or N must be 1 for all dimensions specified by dims
```
Comment (StevenWhitaker) [inline src/regularize/diffs.jl:70]: I wrote the original comment before adding `dims`. So, yes, `N > 1` is only necessary for `N[dims]`.

File: src/regularize/diffs.jl [hunk 3]
```diff
@@ -1,111 +1,197 @@
...
+diff_length(N::Dims, dim::Int) = prod(diff_size(N, dim))
+
+
+# check argument validity
+# dims should be Int or AbstractVector{Int} or Dims
+# but using duck typing for simplicity
+function diff_check(N::Dims, dims)
+    (length(unique(dims)) != length(dims)) &&
+        throw(ArgumentError("non-unique dims $dims"))
+	(dim -> diff_check(N, dim)).(dims) # check each dim
```
Comment (StevenWhitaker) [inline src/regularize/diffs.jl:27]: I don't know if I've seen this style before. I would have done `map(dim -> diff_check(N, dim), dims)` instead. But either way works.

Comment (JeffFessler) [reply src/regularize/diffs.jl]: I had never used it before either and decided to try it and it worked.
But `map` is more readable so I've switched it.

**Pull request merged**